src/vm/diagnosticserver.cpp

```diff
@@ -0,0 +1,128 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "common.h"
+#include "diagnosticserver.h"
+#include "diagnosticsipc.h"
+#include "eventpipe.h"
+#include "eventpipeconfiguration.h"
+#include "processdescriptor.h"
+#include "sampleprofiler.h"
+
+#ifdef FEATURE_PAL
+#include "pal.h"
+#endif // FEATURE_PAL
+
+#ifdef FEATURE_PERFTRACING
+
+static DWORD WINAPI DiagnosticsServerThread(LPVOID /*lpThreadParameter*/)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    const ProcessDescriptor pd = ProcessDescriptor::FromCurrentProcess();
+    IpcStream::DiagnosticsIpc ipc("dotnetcore-diagnostic", pd.m_Pid);
+
+    while (true)
+    {
+        IpcStream *pStream = ipc.Accept();  // FIXME: Ideally this would be something like a std::shared_ptr
+        assert(pStream != nullptr);
+        if (pStream == nullptr)
+        {
+            continue;
+        }
+
+        uint32_t nNumberOfBytesRead = 0;
+        MessageHeader header;
+        bool fSuccess = pStream->Read(&header, sizeof(header), nNumberOfBytesRead);
+        if (!fSuccess || nNumberOfBytesRead != sizeof(header))
```

Comment:
I might be wrong, but I think you have to handle the case where the read() function does not guarantee that it returns all the bytes written from the write() side, one might need to call it repeatedly to get all the bytes.

Reply:
I'll address it in the near future.

---
In reply to: [263559720](https://github.com/dotnet/coreclr/pull/23106#discussion_r263559720) [](ancestors = 263559720)

Reply:
This makes me wonder about the read socket code in IpcStream::Read. The twowaypipe code has a loop in it to read until there are no more bytes. I don't know enough about the socket recv to say for sure.